### PR TITLE
refactor: upgrade to PHP 8.0 with rector

### PR DESCRIPTION
**Pull request by kenjis**
Needs #6922

**Description**
See #6921

Applied rules:
* StrEndsWithRector (https://wiki.php.net/rfc/add_str_starts_with_and_ends_with_functions)
* StrStartsWithRector (https://wiki.php.net/rfc/add_str_starts_with_and_ends_with_functions)
* StrContainsRector (https://externals.io/message/108562 https://github.com/php/php-src/pull/5179)

**Checklist:**
- [x] Securely signed commits
- [ ] Component(s) with PHPDoc blocks, only if necessary or adds value
- [ ] Unit testing, with >80% coverage
- [ ] User guide updated
- [x] Conforms to style guide

### CODE REVIEW

File: system/Autoloader/FileLocator.php [hunk 1]
```diff
@@ -123,7 +125,7 @@ public function getClassname(string $file): string
                 continue;
             }
 
-            if ((isset($tokens[$i - 2][1]) && ($tokens[$i - 2][1] === 'phpnamespace' || $tokens[$i - 2][1] === 'namespace')) || ($dlm && $tokens[$i - 1][0] === T_NS_SEPARATOR && $token[0] === T_STRING)) {
+            if ((isset($tokens[$i - 2][1]) && ($tokens[$i - 2][1] === 'phpnamespace' || $tokens[$i - 2][1] === 'namespace')) || ($dlm && $tokens[$i - 1][0] === T_NS_SEPARATOR && $token->is(T_STRING))) {
```
Comment (kenjis) [inline system/Autoloader/FileLocator.php:128]: @samsonasik 
This code causes an error.

> Running PHPStan...
Note: Using configuration file .../CodeIgniter4/phpstan.neon.dist.
Error thrown in .../CodeIgniter4/system/Autoloader/FileLocator.php on line 128 while loading bootstrap file .../CodeIgniter4/phpstan-bootstrap.php: Cannot use object of type PhpToken as array

Comment (samsonasik) [reply system/Autoloader/FileLocator.php:128]: Possibly bug on TokenGetAllToObjectRector

**Comment by samsonasik** [comment]
Cherry-picking rule by rule per-PR from config

https://github.com/rectorphp/rector-src/blob/2b35d2e75ea03ffe643c012485952ec6df3ea2ba/config/set/php80.php#L34


can be a way to make easier to catch breaking changes, so smallest PR will be easy to merge

File: system/Autoloader/FileLocator.php (hunk 1 above)
Comment (kenjis) [reply system/Autoloader/FileLocator.php:128]: Yes, this is a bug on TokenGetAllToObjectRector. 
The `$tokens` in the original code is just an array of array, and used as `$tokens[$i - 2][1]`.
But `$tokens` in the refactored code is an array of PhpToken object.

File: system/Cookie/Cookie.php [hunk 2]
```diff
@@ -630,7 +626,7 @@ public function toHeaderString(): string
     /**
      * {@inheritDoc}
      */
-    public function __toString()
+    public function __toString(): string
```
Comment (kenjis) [inline system/Cookie/Cookie.php:629]: Is this a breaking change?

Comment (samsonasik) [reply system/Cookie/Cookie.php:629]: No, `__toString` is special, it keep working even child doesn't has return type

**Comment by kenjis** [comment]
It seems difficult to apply ClasPropertyAssignToConstructorPromotionRector.
Some classes will be broken due to the existing type mismatches.

**Comment by samsonasik** [comment]
That can posibly due to autoload overlapped, the tweak may be needed https://github.com/rectorphp/rector/blob/main/docs/static_reflection_and_autoload.md

**Comment by kenjis** [comment]
Probably due to property type mismatch between parent class and child class. 
Fixing the type declaration might result in BC, so I skipped them for now.
See #6924

File: system/Autoloader/FileLocator.php (hunk 1 above)
Comment (kenjis) [reply system/Autoloader/FileLocator.php]: Skipped the rule.

**Comment by MGatner** [comment]
I think the small approach is good. I will also note that I've found CS Fixer to be a great companion tool for Rector during the upgrades to make sure the results are still compliant with our style guide.

**Pull request closed without merging**